editor/src/components/github/github-repository-clone-flow.tsx

```diff
@@ -0,0 +1,179 @@
+import { useOnClickAuthenticateWithGithub } from '../../utils/github-auth'
+import { Dialog, FormButton } from '../../uuiui'
+import { isLoggedIn, type EditorDispatch } from '../editor/action-types'
+import { setGithubState } from '../editor/actions/action-creators'
+import { useDispatch } from '../editor/store/dispatch-context'
+import { type EditorStorePatched, type GithubRepoWithBranch } from '../editor/store/editor-state'
+import { Substores, useEditorState, useRefEditorState } from '../editor/store/store-hook'
+import { onClickSignIn } from '../titlebar/title-bar'
+
+export const LoadActionsDispatched = 'loadActionDispatched'
+
+export const GithubRepositoryCloneFlow = React.memo(() => {
+  const githubRepo = useEditorState(
+    Substores.userState,
+    (store) => store.userState.githubState.gitRepoToLoad,
+    'GithubRepositoryCloneFlow gitRepoToLoad',
+  )
+  const userLoggedIn = useEditorState(
+    Substores.userState,
+    (store) => isLoggedIn(store.userState.loginState),
+    'GithubRepositoryCloneFlow userLoggedIn',
+  )
+
+  const githubAuthenticated = useEditorState(
+    Substores.userState,
+    (store) => store.userState.githubState.authenticated,
+    'GithubRepositoryCloneFlow githubAuthenticated',
+  )
+
+  const onClickAuthenticateWithGithub = useOnClickAuthenticateWithGithub()
+
+  if (githubRepo == null) {
+    // we don't want to load anything, so just return null to hide this overlay
+    return null
+  }
+  if (!userLoggedIn) {
+    // we want to prompt the user to log in
+    return (
+      <Dialog
+        title='Not Signed In'
+        content={<>You need to be signed in to be able to clone a repository from GitHub</>}
+        closeCallback={NO_OP}
+        defaultButton={
+          <FormButton primary onClick={onClickSignIn}>
+            Sign In To Utopia
+          </FormButton>
+        }
+      />
+    )
+  }
+  if (!githubAuthenticated) {
+    // we want to prompt the user to log authenticate their github
+    return (
+      <Dialog
+        title='Connect to Github'
+        content={
+          <>
+            You need to connect Utopia with your Github account to be able to access Github
+            repositories
+          </>
+        }
+        closeCallback={NO_OP}
+        defaultButton={
+          <FormButton primary onClick={onClickAuthenticateWithGithub}>
+            Authenticate with Github
+          </FormButton>
+        }
+      />
+    )
+  }
+
+  // The GitClonePseudoElement triggers the actual repo cloning
+  return <GitClonePseudoElement githubRepo={githubRepo} />
+})
+
+// The git repo clone flow is initiated from the URL, which means we only ever want to do it once per editor load
+let didWeInitiateGitRepoDownloadSinceTheEditorLoaded = false
+
+async function cloneGithubRepo(
+  dispatch: EditorDispatch,
+  storeRef: { current: EditorStorePatched },
+  githubRepo: GithubRepoWithBranch,
+) {
+  if (didWeInitiateGitRepoDownloadSinceTheEditorLoaded) {
+    return
```

Comment:
Should this throw an error instead? This feels like something that shouldn't fail silently

Reply:
this function gets called a couple of times because of ancestor remounts